lib/Interpreter/IncrementalCUDADeviceCompiler.cpp

```diff
@@ -47,46 +47,54 @@ namespace cling {
       llvm::errs() << "Error: CudaGpuBinaryFileNames can't be empty\n";
       return;
     }
 
-    if (!generateHelperFiles()) return;
     if (!findToolchain(invocationOptions)) return;
     setCuArgs(CI.getLangOpts(), invocationOptions, optLevel,
               CI.getCodeGenOpts().getDebugInfo());
 
-    m_HeaderSearchOptions = CI.getHeaderSearchOptsPtr();
-
-    // This code will write to the first .cu-file. It is necessary that some
-    // cling generated code can be handled.
-    const std::string initialCUDADeviceCode =
-        "extern void setValueNoAlloc(void* vpI, void* vpV, void* vpQT, char "
-        "vpOn, float value);\n"
-        "extern void setValueNoAlloc(void* vpI, void* vpV, void* vpQT, char "
-        "vpOn, double value);\n"
-        "extern void setValueNoAlloc(void* vpI, void* vpV, void* vpQT, char "
-        "vpOn, long double value);\n"
-        "extern void setValueNoAlloc(void* vpI, void* vpV, void* vpQT, char "
-        "vpOn, unsigned long long value);\n"
-        "extern void setValueNoAlloc(void* vpI, void* vpV, void* vpQT, char "
-        "vpOn, const void* value);\n";
+    // cling -std=c++xx -Ox -x cuda -S --cuda-gpu-arch=sm_xx --cuda-device-only
+    // ${include headers} ${-I/paths} [-v] [-g] ${m_CuArgs->additionalPtxOpt}
+    std::vector<std::string> argv = {"cling",
+                                     m_CuArgs->cppStdVersion.c_str(),
+                                     m_CuArgs->optLevel.c_str(),
+                                     "-x",
+                                     "cuda",
+                                     "-S",
+                                     m_CuArgs->ptxSmVersion.c_str(),
+                                     "--cuda-device-only"};
+    addHeaderSearchPathFlags(argv, CI.getHeaderSearchOptsPtr());
+    if (m_CuArgs->verbose) argv.push_back("-v");
+    if (m_CuArgs->debug) argv.push_back("-g");
+    argv.insert(argv.end(), m_CuArgs->additionalPtxOpt.begin(),
+                m_CuArgs->additionalPtxOpt.end());
 
-    std::error_code EC;
-    llvm::raw_fd_ostream cuFile(m_FilePath + "cling0.cu", EC,
-                                llvm::sys::fs::F_Text);
-    if (EC) {
-      llvm::errs() << "Could not open " << m_FilePath << "cling0.cu"
-                   << EC.message() << "\n";
+    std::vector<const char*> argvChar;
+    argvChar.resize(argv.size() + 1);
+
+    std::transform(argv.begin(), argv.end(), argvChar.begin(),
+                   [&](const std::string& s) { return s.c_str(); });
+
+    // argv list have to finish with a nullptr.
+    argvChar.push_back(nullptr);
+
+    // create incremental compiler instance
+    m_PTX_interp.reset(new Interpreter(argvChar.size(), argvChar.data()));
+
+    if (!m_PTX_interp) {
+      llvm::errs() << "Could not create PTX interpreter instance\n";
       return;
     }
 
-    cuFile << initialCUDADeviceCode;
-    cuFile.close();
+    // initialize NVPTX backend
+    LLVMInitializeNVPTXTargetInfo();
```

Comment:
CI says:
```
error: ‘LLVMInitializeNVPTXTargetInfo’ was not declared in this scope
     LLVMInitializeNVPTXTargetInfo();
error: ‘LLVMInitializeNVPTXTarget’ was not declared in this scope
     LLVMInitializeNVPTXTarget();
error: ‘LLVMInitializeNVPTXTargetMC’ was not declared in this scope
     LLVMInitializeNVPTXTargetMC();
error: ‘LLVMInitializeNVPTXAsmPrinter’ was not declared in this scope
     LLVMInitializeNVPTXAsmPrinter();
```

missing includes?

Reply:
There is a problem with the Travis CI configuration. Travis builds cling with the cmake argument `-DLLVM_TARGETS_TO_BUILD=host`. So only the x86 backend is built.

Reply:
But that's ok to do. Maybe you want to add guards for the build target?

Reply:
@Axel-Naumann Do you have a better solution? A `#ifdef` guard is just a workaround right now. It's just avoids the test build with the new functionality and that's not a good solutions. I think enabling the nvptx backend is the better solution. It should no affect the build on systems that don't provide a CUDA runtime.

Reply:
I solved the problem by adding the `NVPTX` backend to `-DLLVM_TARGETS_TO_BUILD=host`. It does not cause any problems on non CUDA system. It just increases the size a little.